src/shogun/mathematics/linalg/Vector.cpp

```diff
@@ -0,0 +1,226 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/Vector.h>
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+template<class T>
+Vector<T>::Vector()
+{
+	init();
+}
+
+template<class T>
+Vector<T>::Vector(SGVector<T> const &vector)
+{
+	init();
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector + vector.vlen, m_data.get());
+	m_len = vector.vlen;
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T> const &vector)
+{
+	init();
+
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+									[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+}
+
+template<class T>
+Vector<T>::~Vector()
+{
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data.get());
+	m_len = vector.vlen;
+	m_onGPU = false;
+	m_gpu_impl.release();
+	return *this;
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(Vector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+	 							[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl.reset(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		m_onGPU = false;
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+	return *this;
+}
+
+template<class T>
+Vector<T>::operator SGVector<T>()
+{
+	if (onGPU())
+		transferToCPU();
+	SGVector<T> vector(m_len);
+	std::copy(m_data.get(), m_data.get()+m_len, vector.vector);
+	return vector;
+}
+
+template<class T>
+bool Vector<T>::onGPU() const
+{
+	return m_onGPU;
+}
+
+template<class T>
+T* Vector<T>::data()
+{
+	if (onGPU())
+		transferToCPU();
+	return m_data.get();
+}
+
+template<class T>
+T const * Vector<T>::data() const
+{
+	return m_data.get();
```

Comment:
See, we should not sync the data in everything that we do! GPU data, if present, should be copied to CPU *only when* (a) transferToCPU method is called explicitly or (b) an SGVector is created from this object. Otherwise we'll be ending up doing the transfer all the time! That's not very good!

The point of these `data()` methods were to access the CPU data by CPU backend (say, when you want to create an eigen map, which will only be used when no data is there in the GPU). The GPU backend anyway accesses the gpu data (it should not call these `data()` methods. See what I mean?

Just make the data methods as it was before. In addition to that, let's make onGPU `false` explicitly when non-const `data()` is called. The reason is that, non-const access means that we're modifying it, in which case GPU data should become obsolete.